builtin/logical/pki/path_acme_order.go

```diff
@@ -94,7 +107,403 @@ func patternAcmeGetOrder(b *backend, pattern string) *framework.Path {
+	}
+
+	cert, err := x509.ParseCertificate(certEntry.Value)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing certificate %s: %w", order.CertificateSerialNumber, err)
+	}
+
+	issuer, err := ac.sc.fetchIssuerById(order.IssuerId)
+	if err != nil {
+		return nil, fmt.Errorf("failed loading certificate issuer %s from storage: %w", order.IssuerId, err)
+	}
+
+	allPems, err := func() ([]byte, error) {
+		leafPEM := pem.EncodeToMemory(&pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: cert.Raw,
+		})
+
+		chains := []byte(issuer.Certificate)
+		for _, chainVal := range issuer.CAChain {
+			if chainVal == issuer.Certificate {
+				continue
+			}
+			chains = append(chains, []byte(chainVal)...)
+		}
+
+		return append(leafPEM, chains...), nil
+	}()
+	if err != nil {
+		return nil, fmt.Errorf("failed encoding certificate ca chain: %w", err)
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			logical.HTTPContentType: "application/pem-certificate-chain",
+			logical.HTTPStatusCode:  http.StatusOK,
+			logical.HTTPRawBody:     allPems,
+		},
+	}, nil
+}
+
+func (b *backend) acmeFinalizeOrderHandler(ac *acmeContext, _ *logical.Request, fields *framework.FieldData, uc *jwsCtx, data map[string]interface{}, account *acmeAccount) (*logical.Response, error) {
+	orderId := fields.Get("order_id").(string)
+
+	csr, err := parseCsrFromFinalize(data)
+	if err != nil {
+		return nil, err
+	}
+
+	order, err := b.acmeState.LoadOrder(ac, uc, orderId)
+	if err != nil {
+		return nil, err
+	}
+
+	if order.Status == ACMEOrderPending {
+		// Lets see if we can update our order status to ready if all the authorizations have been completed.
+		if requiredAuthorizationsCompleted(b, ac, uc, order) {
+			order.Status = ACMEOrderReady
+		}
+	}
+
+	if order.Status != ACMEOrderReady {
+		return nil, fmt.Errorf("%w: order is status %s, needs to be in ready state", ErrOrderNotReady, order.Status)
+	}
+
+	if !order.Expires.IsZero() && time.Now().After(order.Expires) {
+		return nil, fmt.Errorf("%w: order %s is expired", ErrMalformed, orderId)
+	}
+
+	if err = validateCsrMatchesOrder(csr, order); err != nil {
+		return nil, err
+	}
+
+	if err = validateCsrNotUsingAccountKey(csr, uc); err != nil {
+		return nil, err
+	}
+
+	signedCertBundle, issuerId, err := issueCertFromCsr(ac, csr)
+	if err != nil {
+		return nil, err
+	}
+
+	hyphenSerialNumber := normalizeSerialFromBigInt(signedCertBundle.Certificate.SerialNumber)
+	err = storeCertificate(ac.sc, signedCertBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	order.Status = ACMEOrderValid
+	order.CertificateSerialNumber = hyphenSerialNumber
+	order.CertificateExpiry = signedCertBundle.Certificate.NotAfter
+	order.IssuerId = issuerId
+
+	err = b.acmeState.SaveOrder(ac, order)
+	if err != nil {
+		b.Logger().Warn("orphaned generated ACME certificate due to error saving order", "serial_number", hyphenSerialNumber, "error", err)
```

Comment:
We could make an attempt at deleting it first if we want? Not sure your thoughts. :-)

Reply:
Yeah, I could try a delete and trap any errors coming from that if storage has died on us completely. Will do it in another PR to get this merged.